File: src/domain/slug.ts
const NON_SLUG_CHARACTERS = /[^\p{Letter}\p{Number}]+/gu;
const LEADING_OR_TRAILING_HYPHENS = /^-+|-+$/g;

function normalizeBaseSlug(value: string): string {
  return value
    .toLowerCase()
    .replace(NON_SLUG_CHARACTERS, "-")
    .replace(LEADING_OR_TRAILING_HYPHENS, "");
}

function pageSuffix(pageId: string): string {
  return pageId.replace(/-/g, "").slice(0, 8);
}

export function createSlug(title: string, pageId?: string): string {
  const slug = normalizeBaseSlug(title);
  if (slug) {
    return slug;
  }

  const suffix = pageId ? pageSuffix(pageId) : "";
  return suffix ? `page-${suffix}` : "page";
}

export function createUniqueSlug(base: string, exists: (slug: string) => boolean, pageId: string): string {
  if (!exists(base)) {
    return base;
  }

  const suffix = pageSuffix(pageId);
  const candidate = suffix ? `${base}-${suffix}` : `${base}-page`;
  if (!exists(candidate)) {
    return candidate;
  }

  let counter = 2;
  while (exists(`${candidate}-${counter}`)) {
    counter += 1;
  }

  return `${candidate}-${counter}`;
}

File: src/main/kotlin/xyz/robinjoon/notionblog/domain/model/Slug.kt
package xyz.robinjoon.notionblog.domain.model

import java.util.Locale

object Slug {
    private val nonSlugCharacters = Regex("[^\\p{L}\\p{N}]+")

    fun fromTitle(title: String, pageId: String? = null): String {
        val slug = title
            .lowercase(Locale.ROOT)
            .replace(nonSlugCharacters, "-")
            .trim('-')

        if (slug.isNotEmpty()) {
            return slug
        }

        val suffix = pageId?.replace("-", "")?.take(8).orEmpty()
        return if (suffix.isEmpty()) "page" else "page-$suffix"
    }

    fun unique(base: String, pageId: String, exists: (String) -> Boolean): String {
        if (!exists(base)) {
            return base
        }

        val suffix = pageId.replace("-", "").take(8)
        val candidate = if (suffix.isEmpty()) "$base-page" else "$base-$suffix"
        if (!exists(candidate)) {
            return candidate
        }

        var counter = 2
        while (exists("$candidate-$counter")) {
            counter += 1
        }
        return "$candidate-$counter"
    }
}
